### ISSUE TITLE
Massive Project64 audit

### ISSUE THREAD

**Issue by vgturtle127** [project64/project64#773]
# Project64 audit and testing

## Introduction

Alright, so initially this started out as a stupid joke I made on IRC (except in that context I was saying I would test every N64 emulator ever, including stuff like TRWin and Nemu64), but I actually think this might be good for not only Project64 and all of the plugins, but Mupen64 (along with it's forks and derivatives) and 1964 (and all of IT'S forks and derivatives). The reason it will benefit them is because if we improve the plugins, the emulators might work way better with more stable, accurate plugins. Or not, I don't really know. I haven't looked at the 1964 or Mupen64 code base in a long time. :confused: 

The only eligible emulators and plugins should be open-source, otherwise we can't improve them, or do anything else for that matter. That said, I may make exceptions for certain plugins if there is enough demand for it (such as Jabo's plugins for graphics, or Shunyuan's plugins). I will also list depreciated, old, or obscure plugins, and we can discuss whether or not we will test them or not. I don't anticipate on testing many input plugins, and most of the older and more obscure plugins will either not work, or are extremely basic and don't emulate much. Some also use outdated software or outdated plugin specifications which might not work with newer emulators.

Basically, I want to test Project64 (just Project64 for now) along with every combination of plugin that would make sense. While I am doing this, I will of course document any bugs, glitches, inaccurate behaviors, etc. I will then try to generate reports based on all my findings, and of course report all issues to the authors and send pull requests/issues as well (if applicable).

So initially, I will be testing Projec64 32-bit and 64-bit. If Project64 2.3 comes out soon, I will test that version, if not I will test 2.2 and the latest compiled from source.

I can test Project64 1.6 in addition if you guys want, but ideally, I'd really rather not. :laughing: 

Mainly because it is the old version, and sources for it haven't been released I don't think. I know sources for 1.4 and some plugins and some other random stuff has been released, but not any 1.6/1.6.1 stuff. Not even 1.7, if I remember correctly.

As for what plugins to test, I will make a list below of all known plugins. This should be an exhaustive list, but if you know of any other plugins or know information that has a question mark next to it, I would appreciate the information! If it has a question mark, I either don't know, need to do some research, or it is up for discussion. As usual, I am open for feedback or discussion.

Marked plugins are plugins I will definitely be testing.

# Plugins

## Graphics plugins

* [x] Jabo's Direct3D 1.7.0.57-ver5 (other versions possibly?) [closed-source?]
* [x] Glide64 final/2.0.0.1 (other versions possibly?) [open-source]
* [x] GlideN64 (stable and latest compiled from source?) [open-source]
* [x] Shunyuan's "SoftGraphic" (version 1.5) [closed-source]
* [x] Rice video (0.4.4 and latest compiled from source?) [open-source]
* [x] Z64/Z64GL (r17 and Angrylion's modified r17 build) [open-source]
* [x] HatCat's/Angrylion's pixel-accurate OpenGL plugin (June 13th, 2015 build) [open-source]
* [ ] glN64 (versions?) [open-source]
* [ ] Direct64 (version?) [closed-source]
* [ ] 1964 OpenGL plugin (version?) [open-source]
* [ ] Zilmar's CFB plugin (version 1.0) [open-source]
* [ ] RCP's Direct3D plugin (version?) [closed-source]
* [ ] Arachnoid plugin (version 1.0) [open-source]
* [ ] TR64 Direct3D (version?) [open-source]
* [ ] Icepir8's TR64 OpenGL (version?) [open-source]
* [ ] More?

## Audio plugins

* [x] Jabo's DirectSound plugin (requires an RSP plugin) (version?) [closed-source?]
* [x] Azimer's HLE audio plugin (versions 0.56 WIP 2, 0.60 WIP 2, 0.70 WIP 4, latest from source?) [open-source]
* [x] Shunyuan's HLE audio plugin (version 1.8.2u1) [open-source]
* [ ] Zilmar's basic audio plugin (version 1.0) [open-source]
* [ ] Zilmar's no sound plugin (version 1.0) [open-source]
* [ ] TR64 sound plugin (version?) (source?)
* [ ] More?

## Input plugins

* [x] Jabo's DirectInput (version?) [closed-source]
* [ ] Zilmar's basic keyboard plugin (version 1.0) [open-source]
* [x] N-Rage (latest stable and compiled from source?) [open-source]
* [ ] DarkMan's DirectInput plugin (version?) [open-source]
* [ ] Def's DirectInput plugin (version?) [open-source]
* [ ] NooTe's DirectInput plugin (version?) [open-source]
* [ ] TR64 input plugin (version?) [closed-source?]
* [ ] Blight's input plugin (version?) [open-source]
* [ ] Billard's input plugin (version?) [closed-source]
* [ ] More?

## RSP plugins

* [x] Zilmar's RSP (version?) [open-source]
* [x] Mupen64 HLE RSP (version?) [open-source]
* [x] Z64 plugins (Ziggy-z64RSP, Ziggy-PJ64, Z64-Angrylion) (versions?) [open-source]
* [x] HatCat's RSP interpreter (versions?) [open-source]
* [ ] More?

Have suggestions? Comments? Concerns? Do you have source or binaries for plugins listed or otherwise that are not easy to find or not released yet? Please get in touch!

:octocat: :+1: :smiley:

**Comment by toehead2001** [comment]
Sounds like you have a long road ahead of you. Are you looking to do it all yourself? I know *Bighead* did the same thing at one point. Perhaps you can contact him, and see if he's interested in doing it again in a collaboration. His email address is on his site. http://bhemuhelp.co.nf/

It would be nice if you published updates to a public Google Spreadsheet, or something similar.

**Comment by AmbientMalice** [comment]
This is a rather complicated undertaking. A great many of those plugins are either dead or redundant. Even the open source ones. It's not a simple matter to audit these plugins if you're not already aware of their problems, and why they have those problems. I do admire your passion, of course.

**Comment by LegendOfDragoon** [comment]
@vgturtle127 You're missing 1964 audio, which is open source. Also, Shunyuan's audio is not open source. The "source" he provided is not only outdated / different from latest binary, but it also was not the complete source.

**Comment by cxd4** [comment]
It was, however, a tiny little fork of 1964 Audio. :)

**Comment by purplemarshmallow** [comment]
You are missing Lemmy's gfx plugin https://github.com/purplemarshmallow/Nemu64
And Pokopom input https://github.com/KrossX/Pokopom

**Comment by weinerschnitzel** [comment]
You should only bother testing the active ones now (PJ642.X, M64P, GLideN64, AziAudio, HatCat's RSP, Bsmiles RSP, PJ64 RSP), cross reference the stable ones (Glide64, PJ64 1.6, M64 0.5, 1964 1.1) and only use the most obscure ones like Lemmy's for their unique advantages. I.E. limited Rogue Squadron HLE.

Stay away from forks, unless it ONLY has known advantages. While I think it would be nice to include Rice-Video in whatever database you are building, the current Git for Rice-Video has some known regressions from original Rice 6.11 and 1964video 1.0.0.

I love the 1964 emulator and plugins, but I do feel time can be better spent in the active projects and their goals.

**Comment by LegendOfDragoon** [comment]
>I love the 1964 emulator and plugins, but I do feel time can be better spent in the active projects and their goals.

I think 1964 is useful and worth testing.

>Stay away from forks, unless it ONLY has known advantages.

I agree with this though.

**Comment by weinerschnitzel** [comment]
I should have been clearer. 1964 1.1 would be useful to reference as a known stable emulator, but the associated plugins have competitors with better results and approaches. For example, audio HLE is already fully attainable in other projects like Bsmiles' RSP. Comparing 1964audio for speed is good, but I suppose the the problems with the plugin are not going to be easily solvable for that approach without being able to examine the RSP decompiler source. 

AziAudio 0.7 WIP5 and WIP6 could use more diff testing. Thoroughly testing both of those would be beneficial for that project.

I would like it if this went a step further than a compatibility list, and associated every little issue with every plugin combination.

**Comment by LegendOfDragoon** [comment]
>1964 1.1 would be useful to reference as a known stable emulator, but the associated plugins have competitors with better results and approaches. For example, audio HLE is already fully attainable in other projects like Bsmiles' RSP

I still think the plugins are useful. I actually use a tweaked version of 1964 audio for certain games like WDC, Stunt Racer, Rogue Squadron (sometimes), and Top Gear Rally. For audio, I'm more concerned about AI emulation, latency, and buffer handling, because the sound quality and latency are more important to me than a minor performance boost. I understand you are interested in N64 emulation on XBOX so performance matters more to you. I recommend writing your own optimized algorithms (using mmx2 and sse1) to get better performance.

Even Rice Video has its uses. I still learn things from it, from time to time. I will probably study it even more, when I finally decide to work on a hardware rendering graphics plugin. It's understandable to prioritize Jabo and Glide64 when it comes to testing, but I think it's still a good idea to at least do some testing with Rice Video.

@weinerschnitzel you mind uploading your zilmar spec port of Bsmile's RSP? I did a sloppy port over a year ago and encountered an issue (I'm wondering if it's present in the up to date code). Basically Body Harvest's audio sounds messed up, compared to LLE audio.

**Comment by vgturtle127** [comment]
@toehead2001 Yeah, I will definitely do something like that. Probably on GitHub, though. I will get a hold of Bighead and see what he wants to 

@AmbientMalice Yeah, some of the plugins I probably won't test. Especially the closed-source plugins. GlideN64 is an interesting beast, but one I am definitely going to tackle regardless, I like a challenge anyway. I once re-wrote a PS1 plugin like 8 or 9 times to just get one game working. I got this. :smile: 

@LegendOfDragoon I updated the post with your suggestions! Thanks!

@cxd4 Huh, I didn't know that.

@purplemarshmallow Added Pokopom and Lemmy's graphics plugin, and I will try to get Lemmy's plugin compiled or something to see if it works in any useful capacity. Thanks!

@weinerschnitzel Yeah, mostly active and open-source stuff will be tested. Stable plugins and a few closed-source plugins (Jabo's) will be tested as well for reference. All the forks will be scrutinized appropriately. If they have any advantages, they will be documented. Thanks!

@LegendOfDragoon I think I will test Mupen64 and 1964 and their forks and stuff independently. One emulator at a time. Mupen64Plus will be weird because different plugins will work for each version of it and Mupen64, but again, I'm not a person who is afraid to do what needs to be done. Actually, 1964 might come after since it uses the same plugin specifications and could be easily upgraded to the new spec. I'll cross that bridge when I come to it, though. :smile: 

@weinerschnitzel I think 1964 1.1 and 1.2 should both be tested. At least, I think so. Last time I checked the code, anyway. I thought there was a big difference between the two. Maybe test 0.9.9 and 0.8.5 as well? Again, I will go deeper into it after I test Project64.

By the way, I will be using a USB N64 controller for all of this. I thought that was worth mentioning, I also added it above. I also added my PC specs. It is pretty budget, but modern. So, in the end it doesn't matter so much, since it is only N64 emulation. Although, I suppose that is a silly thing to say considering where we are. :smile: 

@LegendOfDragoon Do I need to add Bsmile's RSP to the list?

**Comment by vgturtle127** [comment]
@purplemarshmallow Is the gs_dll.dll the graphics plugin? And does it use the Zilmar spec?

**Comment by purplemarshmallow** [comment]
You can download a binary here https://github.com/purplemarshmallow/Nemu64/releases
It's named lemd39.dll and works with PJ64. gs_dll is Nemu's cheat module

**Comment by cxd4** [comment]
@purplemarshmallow so was it you that made a Direct3D9 fork of Lemmy's Nemu graphics plugin, or was it somebody else who changed it to D3D9?

**Comment by purplemarshmallow** [comment]
Lemmy did that. His original plugin bundled with Nemu 0.8 used Direct3D8. But later he started working on the plugin again. He used D3D9 and pixelshaders to emulate the combiner. I improved microcode detection and fixed the window and fullscreen code so the plugin also works properly with PJ64 and other emulators.

**Comment by cxd4** [comment]
Doh!  Didn't think that Lemmy's source release would have an update to Direct3D9.

Yeah I don't think he ever released a binary with D3D9 but it looks like he migrated by the time someone asked him for the source.  (So basically there is no D3D8 source.)

**Comment by purplemarshmallow** [comment]
Lemmy released binaries with D3D9 but he did not bother to change the name it still said D3D8. I renamed the plugin.

**Comment by cxd4** [comment]
Must have missed it?

I have his LemD3DSource.zip on disk from years ago but it is source-only, no binaries.

**Comment by purplemarshmallow** [comment]
He released it here but it got removed
http://www.emutalk.net/threads/21092-This-will-allow-you-to-export-geometry-WITH-textures

**Comment by weinerschnitzel** [comment]
1964 1.2 has some better features, but I think the released binary for 1.2 has regressions. Specifically, RetroRalph's OC changes force CounterFactor to 1 and removes any ability to modify it. Deathdroid changed it later to support CF again, but those revisions came after the 1.2 release. IIRC Super Smash Bros. Vs. Mode requires a CF of 3 to finish a match, and I think DK64 doesn't like CF 1 either.

1.2 also has an RTC patch thanks to Shapyi @ Emutalk.

I think we should work on a new fork of 1964 starting from the 1.1 base, and reapply the beneficial changes from 1.2. I feel safer doing this, rather than picking up straight from Deathdroid's repository. Is there anyone with a strong attention to detail that would mind maintaining it? I feel like I can be sloppy. :/

@LegendOfDragoon 1964audio does scale better and is faster than using Bsmiles RSP in some cases. My issue is of its more internal problems that seem to be unsolvable. You're right though, AI emulation and buffer handling are things that could be referenced for AziAudio. I'll report back when I find time to start a repository for the zilmar spec Bsmiles' RSP. Although, my changes were very trivial, I doubt I got any further than you. I never played Body Harvest though.

**Comment by LegendOfDragoon** [comment]
I have my own fork of 1964, although it has some unofficial changes. I do not think super smash bros requires cf3. I usually use cf1 for most games including SSB. I've never seen a case where 1964 audio's HLE is faster than bsmiles HLE code. So I'm very surprised by your statement. Which game was 1964 audio faster in?  

When you get the chance, please do test body harvest with bsmiles rsp. What internal problems do you have with 1964 audio? I think I fixed most issues with 1964 audio, except for AI emulation, in my experimental build.

**Comment by vgturtle127** [comment]
@purplemarshmallow Thanks, dude! I assumed the gs_dll.dll was a graphics plugin because I thought it was an acronym for graphics software or something. I have never used this plugin before, so I hope it works at least somewhat.

@cxd4 If somebody forked Nemu's graphics plugin, they probably renamed it since Nemu64 wasn't exactly the most compatible or popular plugin. So you may find that 1964 OpenGL or something similar might use some code from it or something. I'm just speculating though.

Does anybody know where the latest Zilmar spec can be found at anytime? Like is there a place where it is kept up to date? A repository that holds it or something? All I can find is version 1.1 of the audio, input, and RSP plugin specs and 1.3 for the graphics spec.

**Comment by project64** [comment]
@vgturtle127 with regards to my specs .. there use to a group (yahoo I believe) where I kept things updated, long before the days of public repositories.

after that they were released in the project64 installer, but they are not anymore. Since they are no longer in the installer and linked to anything they actually got removed out of the repository (https://github.com/project64/project64/commit/9059e307daedaf7bfd8b64221ebc62e1128e6dac)

**Comment by LegendOfDragoon** [comment]
>Do I need to add Bsmile's RSP to the list?

@vgturtle127 I would wait until a zilmar spec port is publicly available.

>@weinerschnitzel @LegendOfDragoon Is Body Harvest a difficult game to run? If so, I should run it extensively during my testing!

Kinda. The audio is bad with many emulators & audio plugins. Then there's also that clipping problem with many graphics plugins. Also cpu recompiler is broken in 1964 1.1 for that game, but not 0.85 (afaik). I wish I could fix some of the regressions in 1964 1.1. Maybe @weinerschnitzel can help me with that :smile: . I mostly just test the game for audio (both microcode and AI emulation).

**Comment by vgturtle127** [comment]
@project64 OK, so I do have the latest versions. Alright, that's all I really wanted to know.

I am going to test every game, so it will be tested regardless. But I want some benchmarks and difficult games to test as well. I am also using homebrew and some SDK stuff, it should help round out the testing.

**Comment by BigheadSMZ** [comment]
Just to throw in my 2 cents, this is indeed a big undertaking. I haven't really tinkered with N64 emulation since the beginning of 2012, so I have no idea what works better now compared to then. Looking back at the 'changelog' ( http://bhemuhelp.co.nf/n64mgcl/source/changes.txt ) for my list, it took nearly 2 months to 'test' every game. And at the time there were far fewer things to work with, and I was unemployed and lacked other hobbies so it was usually an all day thing. :) Now days, time is not on my side...

It might be a good idea to target games that are known to have issues rather than ones that are known to work with multiple configurations (like Mario 64 and Wave Race for example). A few that come to mind are games like Battlezone, Vigilante games, Body Harvest, Top Gear, WCW games, Taz Express, Mia Hamm (and it's incarnates), Gauntlet Legends, F-1 racing games, Star Wars games, Knife Edge (was always too fast), and the many others I'm missing.

**Comment by cxd4** [comment]
@vgturtle127 , bro do u even GameShark ? That's what the gs in gs_dll means.

>So you may find that 1964 OpenGL or something similar might use some code from it or something. I'm just speculating though.

It would be a more sensible speculation if:

* 1964 OpenGL was open-source
* Lemmy's Direct3D plugin was open-source, which it was not until a decade after 1964 OpenGL

**Comment by vgturtle127** [comment]
@BigheadSMZ Hey, Bighead! I got your email. Thanks for all the info. I went ahead and downloaded everything I could (except for the game list, there isn't a "view all at once" option). Problem games will definitely be a focus of mine. The games you mentioned along with Winback Operations (that damn grey square) and a few of the popular titles like Mario Tennis, Ocarina of Time, Yoshi's Story, and more will all be tested because surprisingly, a lot of those games have issues that most people don't notice. For instance, we thought that Mario Tennis had been emulated pretty well until GlideN64 came along. I suppose some configurations might have rendered it well enough, but mostly those were unrealistic configurations at the time.

Ocarina of Time/Majora's Mask STILL have issues. Which sucks because I really like those games and I have been wanting to play through them again. :disappointed: 

I copied the list of games that benefit from overclocking. This seems like it would be cool to add to Project64 at some point. We have seen the benefits of overclocking the CPU and adding additional RDRAM, so I don't see why not. Thanks, Bighead!

Ironically enough, even though I am 24, I am in the same situation as you were when you started this. I am unemployed (recently laid off), and I have other hobbies, but my favorite thing to do is test and work on emulators. So I thought, while I am out of work, why not help the community. :smile: 

@cxd4 Oh, I always thought the GameShark and similar things were never released on the N64. I had a PS1 as a kid, and we always had the Action Replay and GameShark. Good to know, might be worth testing the ROM, now that I think of it!

I didn't know 1964 OpenGL was closed-source. I will modify my list. Thanks for the info! :+1:

**Comment by LegendOfDragoon** [comment]
>It might be a good idea to target games that are known to have issues rather than ones that are known to work with multiple configurations (like Mario 64 and Wave Race for example). A few that come to mind are games like Battlezone, Vigilante games, Body Harvest, Top Gear, WCW games, Taz Express, Mia Hamm (and it's incarnates), Gauntlet Legends, F-1 racing games, Star Wars games, Knife Edge (was always too fast), and the many others I'm missing.

I absolutely agree with this. I feel like too many people test the same few games. Some games I'd add to that list are WDC, Stunt Racer 64, Mortal Kombat 4, Mystical Ninja Starring Goemon, SD Hiryuu no Ken Densetsu, Turok series, Last Legion, Dark Rift, and Bottom of the 9th.

@vgturtle127 I do not think Mupen64 HLE RSP needs much testing because its main issues are known. Also why test Shunyuan's soft graphic plugin? It's an inferior fork of Angrylion's RDP, that is closed source and will likely never improve. Honestly, even his audio plugin is more worthy of testing.

**Comment by vgturtle127** [comment]
@LegendOfDragoon All RSP plugins will be tested mainly for compatibility with other plugins, and to see which plugins have issues with z64 or other plugins that require an RSP plugin. At least, that was my reasoning. Maybe that is something we should all discuss?

Those games are hard to run! I will be sure to keep a list of these games that you guys want tested more extensively. Other than the popular games, I have a pretty good list so far. Thanks guys! :smile: 

As for why I am testing Shunyuan's SoftGraphic plugin, I have heard people say that they have had better luck with it for certain games with Azimer's latest audio plugin compiled form source. Other than that, no reason really. I find it kind of a unique specimen as well, I guess. :hear_no_evil: 

His audio plugin is only slightly a different story. I mean, it runs most things OK, but I find it putting some excess background noise into some games. Maybe it is just my setup, I will have to see during my testing.

**Comment by LegendOfDragoon** [comment]
I forgot to mention that z64gl should be tested! I'd consider trying to improve the plugin, if I could figure out how to fix the driver-specific issues I have with that plugin.

**Comment by vgturtle127** [comment]
@LegendOfDragoon z64 is pretty interesting. What issues are you having? Like what GPU do you have, are you running Windows, what version, what version of z64 are you using, etc?

**Comment by LegendOfDragoon** [comment]
I have an Intel IGP, so on Windows I have additional issues with z64gl than I do when I use Mesa.

**Comment by purplemarshmallow** [comment]
z64gl is a nice plugin I like it. But it has this strange bug where textures break until you switch fullscreen/window. Does not seem to be driver related I tested on AMD and Nvidia. I think it's fixed in mupen64plus-video-z64. Would be nice to have this fix in the zilmar spec version but I haven't found yet what's wrong.

But I managed to compile z64gl it with Visual studio and I have the project files anyone intersted?

**Comment by cxd4** [comment]
Yeah I know what you mean.  The textures crash every random, undefined amount of minutes/seconds.  Switching between windowed and full-screen with Alt+Enter is my getaround with z64gl for that too.  I believe it's more of a TMEM buffer overrun than a driver issue or OpenGL implementation defect but don't really know either.

**Comment by LegendOfDragoon** [comment]
>But I managed to compile z64gl it with Visual studio and I have the project files anyone intersted?

I'm interested :smile: .

>I think it's fixed in mupen64plus-video-z64. Would be nice to have this fix in the zilmar spec version but I haven't found yet what's wrong.

That's interesting! I didn't know their version fixed any bugs. Hopefully it's documented in their commit history.

**Comment by cxd4** [comment]
Likely not.  I'd say it was an accidental fix, arising from philosophical cleanups.

**Comment by vgturtle127** [comment]
@LegendOfDragoon You use integrated graphics? No emulation for you! :smile: 

I can literally get nothing to work with integrated stuff. Emulators, games, sometimes even software. I don't know how some people live on integrated graphics. Even a sub-100$ graphics card seems to do better for me. Maybe I just suck at getting stuff to work? :laughing: 

@purplemarshmallow I am interested as well!

@cxd4 I will keep this in mind while testing, last time I tried z64 it wouldn't load any game made by Nintendo or Rare without going into this creepy inverted color model mode. I was genuinely creeped out by it.

@cxd4 I've noticed nothing in Git is documented in most repositories I frequent. :laughing: :sob:

**Comment by purplemarshmallow** [comment]
look here https://github.com/purplemarshmallow/z64
You will need dependencies: glew and sdl. I might add the dependencies to the repo later
You will need to put glew32.dll and SDL.dll in the emulator's directory

**Comment by vgturtle127** [comment]
@purplemarshmallow Would you mind adding a binary for the plugin to releases? I don't currently have Visual Studio installed, and I can't until I free some space on my SSD. I would pay you in grey squares from WinBack...

**Comment by vgturtle127** [comment]
Hey, everybody. I updated a ton of stuff in my original post. Tell me if I said anything stupid or made any mistakes or anything. :smile:

**Comment by LegendOfDragoon** [comment]
>@LegendOfDragoon You use integrated graphics? No emulation for you! :smile:
I can literally get nothing to work with integrated stuff. Emulators, games, sometimes even software. I don't know how some people live on integrated graphics. Even a sub-100$ graphics card seems to do better for me. Maybe I just suck at getting stuff to work? :laughing:

IGPs really do suck ;/ . I get tons of problems that no one else seems to have, like stack hash errors with Rice Video when using D3D. Graphics cards make a huge difference!

@purplemarshmallow thanks for uploading the repo, I'll be sure to test it out soon. I wonder why it requires SDL.dll when the regular z64gl.dll binary does not.

**Comment by vgturtle127** [comment]
@LegendOfDragoon I wondered that too...

**Comment by cxd4** [comment]
IIRC z64 and z64gl (and angrylion fork of rsp) as well as other MAME-derived plugin sources generally used SDL functions and methods.  ziggy probably just had a build-time define in the make system that affected the build so SDL wasn't needed whenever he built it.

**Comment by vgturtle127** [comment]
@cxd4 That means nobody should need it, right?

I also added some more updates before I go to bed tonight. Tomorrow I am going to try and finish my research and start rounding up the emulators and plugins. After that, I have to get all my ROMs extracted and make sure I have all of them. :smile:

**Comment by weinerschnitzel** [comment]
@LegendOfDragoon It looks like Body Harvest is set to use Link4KBlocks in 1964 1.1 and not in 1964 0.8.5. I did come across a lesser known regression in Link4KBlocks where Conker's Bad Fur Day stopped working with it. 1964 1.1 is configured not to use it, but it is something I needed to do to support ROM paging. You can take a look at my ROM paging fork, the 0.8.5 style LinkBlocks is currently setup as a hack for conker.

https://www.dropbox.com/s/3gtxyv42fghu8up/1964_11_RP.7z?dl=0

**Comment by cxd4** [comment]
@vgturtle127 Nobody ever really "needs" SDL for any program or any plugin.

It's not something one ever "needs" to make programs to require it to run, but the dependency is sometimes there anyway.

For example, let's say somebody on a Macintosh is trying to get rid of the SDL dependency.  They would replace the window and GL context events with Apple's native calls to successfully, easily get rid of it--but then there is one question, how is it going to be ported to Windows if you do that?  Furthermore, how is it going to be ported to Linux?  That's the kind of fundamental code bloat scenario that high-level dependencies (such as the infamous SDL, for a notorious example) generally exist to "solve".

**Comment by vgturtle127** [comment]
@cxd4 Well, I just meant does anybody need it for compiling in this particular instance. But yeah, I get why anybody would use SDL. I have used it in the past, although I mostly use 1.2 for portability reasons.

@weinerschnitzel Link4KBlocks?

Also, when you say ROM paging, what exactly do you mean?

**Comment by cxd4** [comment]
I'm sure there is a way to compile without it, but I have not tried it.  I think angrylion's LOD mod patch of z64gl for a Boss Games issue also had SDL linked into it.  If you pay close attention to the macro logic being used throughout the source I'm sure you'll find the switches necessary to, for Win32, compile without it.

>I mostly use 1.2 for portability reasons.

Well, besides the fact that SDL 2.0+ is ridiculously not at all backwards compatible with SDL 1.x and practically breaks the high-level portability advantage SDL could have conceivably kept in the first place, why else is it you prefer 1.2?

**Comment by vgturtle127** [comment]
Mainly Dreamcast and GP2X and some other systems. Also, apparently SDL2 doesn't support Amiga. That might not sound like a big deal, but I actually have a few emulators and some software I work on that still is compatible with Amiga. So, I like to test that every once in a while.

Basically, SDL2 dropped support for a ton of systems I still am actively developing on. Granted, none of the software is coded by me or anything, but still I like to test it and have the ability to tell people it still works.

**Comment by weinerschnitzel** [comment]
ROM paging is a feature used in XBOX N64 emu's to support reading ROM data from the HDD instead of caching the whole thing to memory. It is required for most games to run on XBOX's with the standard 64MB of RAM. 

Link 4K Blocks is a feature in 1964's recompiler to handle jumps and branches, from 1964 1.1 help manual:
>The purpose of jumps and branches in MIPS code is to jump to a different area of code. The easiest way to emulate one of these jumps in native machine code is to do it indirectly by first returning to the main compiler loop and then by calling the machine code target location from within this loop. This way, the compiler can determine if the target code exists first. If it does not yet exist, meaning that the target has not yet been compiled, it will first compile the code and then call it. This is exactly what happens when Link Code Blocks is set to No.
>
>Of course, it is slow to return to the compiler every time a jump or a branch is encountered. The solution is what I refer to as Just-In-Time linking. The first time a jump is encountered, it is set to jump to a compiler loop. Then when the target code is compiled, if it doesn't already exist, the jump instruction is rewritten to directly jump to the start address of the target machine code. In other words, the first time a jump is encountered, it jumps to the compiler.  The next time the same jump is encountered, it jumps directly to its target. This is what happens when Link Code Blocks is set to Yes. Yes is the default value.  One condition is that only those jump targets within a 4KB aligned boundary can be linked.  Other targets will return to the compiler.

**Comment by vgturtle127** [comment]
@weinerschnitzel Alright, I was just curious. Sounds fancy. :smile: 

Quick question guys. Is 1964 UltraFast the only emulator with overclocking support? And how hard would it be to add this functionality to Project64? It seems like in 1964 it was practically meant to be there. The games run way better, no matter what plugins you use. GoldenEye runs at like four times the speed. It still only outputs to 60 VI's or whatever, but the extra speed is a drastic difference. And with a GameShark cheat, you can also add an unlocked frame rate. :smile:

**Comment by Frank-74** [comment]
1964 1.2 has overclocking support as well.  And I agree it makes a massive difference.  Especially on my netbook, which has a poor 1.4Ghz quad core Atom processor.  Goldeneye runs like silk.

**Comment by LegendOfDragoon** [comment]
>Is 1964 UltraFast the only emulator with overclocking support?

Yes, along with deathdroid's fork and mine too.

>And how hard would it be to add this functionality to Project64?

Looks easy to do.

@weinerschnitzel I think the body harvest issue is something else. It's slightly less buggy when i disable link blocks, but still unplayable because input is still broken. I wish I was more familiar with 1964's cpu recompiler.

**Comment by Frank-74** [comment]
What causes the clipping issues with Body Harvest?  Walking through buildings and stairs/bridges.

**Comment by vgturtle127** [comment]
@Frank-74 
> netbook
> poor
> 1.4Ghz
> Atom
> ...
> N64 emulation

NOPE. :laughing: 

Oh, wait you said it ran like silk on that? Why is it N64 emulators seem to run OK on Android, Raspberry Pi, and stuff like that, but on desktop computers built yesterday with 1TB of RAM and 14 graphics cards in SLI they run like burnt ass? :laughing: 

@LegendOfDragoon Alright, good to know. Another thing I want to look into. I'm not sure I want to be the one to add support though, otherwise it will be a horrible implementation and we will just have to do it over. :smiley:

**Comment by purplemarshmallow** [comment]
>What causes the clipping issues with Body Harvest? Walking through buildings and stairs/bridges.

I think it needs accurate emulation of auxiliary buffers. I did a lot of testing with GLideN64. You can play it with Jabo's plugin. Jabo's plugin emultes auxiliaty buffers in software it will work. But Jabo's software renderer is not perfect. Buffer creation with RDRAM content needs to be impemented in GLideN64. Then it should work.

**Comment by Frank-74** [comment]
Just tried Jabo with software renderer.  Still walking through stairs.

**Comment by purplemarshmallow** [comment]
You don't need to enable software rendering. Even if it's disabled Jabo's plugin will still use the software renderer for auxiliary buffers. It runs in the background. But it does not work perfectly in Body Harvest you can still walk through stairs

**Comment by Frank-74** [comment]
@purplemarshmallow In Perfect Dark, the light coronas work when you switch on Software Rendering.  And they stay on when you go back into graphics options and turn Software Rendering off again, but now in hires and accelerated.  So its a bit wonky.

**Comment by vgturtle127** [comment]
@purplemarshmallow @Frank-74 Do the plugins that support software rendering (Jabo's does, but what others?) support any kind of logging or graphics dumping of any kind? Besides texture dumping?

**Comment by LegendOfDragoon** [comment]
@vgturtle127 , Angrylion's can log some info. You have to compile the source yourself. Can't remember what was needed to enable the debugging features. Maybe some #defines.

**Comment by purplemarshmallow** [comment]
I updated my z64 repo. It now has separate projects for z64gl and for z64, the software LLE plugin. The dependencies are included everything builds out of the box. I also uploaded binaries https://github.com/purplemarshmallow/z64/releases. I still need to get Ziggy's gfx debugger working, it needs GLFW.

**Comment by vgturtle127** [comment]
@LegendOfDragoon Alright, I will look into that. Thanks!

@purplemarshmallow Thank you very much! I added your build of the plugin to the list and I will be using it for testing!

**Comment by Frank-74** [comment]
@purplemarshmallow 
Just tried your release zip.  Put glew32.dll and SDL.dll in PJ64 directory, z64 and z64gl in plugins directory.  Plugins aren't showing up here.

The original versions work though.

**Comment by vgturtle127** [comment]
@Frank-74 I will test it after I eat dinner. :smile: 

By the way guys, Angrylion has been updating his pixel perfect RDP, and he never posted binaries. So I will be adding the latest binary and testing that. However, there is no makefile or anything so I'm not sure how he is compiling it. I will have to check.

Basically, we are like 8 builds behind in terms of the plugin. He has fixed some issues and improved accuracy. So we should test that instead of the other build on EmuTalk.

**Comment by purplemarshmallow** [comment]
No vs2013. Not sure why it's not showing up for you

**Comment by Frank-74** [comment]
Just checked the files in my System32 directory.  My VS120 files are named, msvcr120_clr0400 and msvcp120_clr0400.  Probably the reason.  I copied them to PJ directory and removed the _clr0400 from the filenames, and now z64 is showing up, but not z64gl.

So I probably need to download the vcredist for 2013.

**Comment by LegendOfDragoon** [comment]
>By the way guys, Angrylion has been updating his pixel perfect RDP, and he never posted binaries. So I will be adding the latest binary and testing that. However, there is no makefile or anything so I'm not sure how he is compiling it. I will have to check.

You can compile with MSVC.

**Comment by vgturtle127** [comment]
@LegendOfDragoon I assume you mean just use the build command on the .cpp files? Why doesn't he have a makefile or solution file?

**Comment by LegendOfDragoon** [comment]
>I assume you mean just use the build command on the .cpp files? Why doesn't he have a makefile or solution file?

I actually never tried downloading from his current repo. I didn't realize he didn't include the solution file. You can download his old solution file, I imagine it still works.

>Do you know if it will compile with Visual Studio Community 2015?

I don't see why not.

**Comment by vgturtle127** [comment]
@LegendOfDragoon Where can I get that? His latest releases on EmuTalk are just binaries and the SourceForge repository doesn't have the solution file in the entire history of the repo...

**Comment by cxd4** [comment]
I, too, am curious to hear of this old solution file of his.

**Comment by LegendOfDragoon** [comment]
My bad, I guess there was no solution file in the older repo either. I really cannot remember how I set it up then. Maybe I got mixed up with other people's builds such as the one mudlord posted.

**Comment by vgturtle127** [comment]
So what do we do if we want to compile it, then? There are clearly MSVC-related things in some of the files, so do we just ask him? :laughing:

**Comment by cxd4** [comment]
I will be getting rid of the solution files once the MSVC-biased optimizations are all neutralized.

It's not that I am against people using MSVC to compile, or why would I have created the Visual Studio solution for people to use in the first place.  But I prefer having fewer build files, more source files.

>EDIT: I forgot I have no room for Visual Studio on my SSD.

You're "supposed" to compile it with MSVC 2002 like he does (with or without Visual Studio), but I think he's tested MinGW from time to time as well or it would not have been able to compile it out of the box.

**Comment by vgturtle127** [comment]
@cxd4 Did you find them? The solution files, I mean?

**Comment by cxd4** [comment]
No, I created them.

That was a very long time ago, back before I had even split up the source into multiple files.

**Comment by cxd4** [comment]
It's not possible to have VC solution files for VS2003 and earlier, anyway, so it stands to reason that any conjecture of them existing before I created them is futile.  Now, mudlord, I can't remember if he used the ones from my old (and deleted, ancient) fork from way back then, or made his own just like I did.  But there is no such thing as a Visual Studio solution file for VC .NET 2002 which of course is what angrylion uses.

I recommend using whatever's lightweight and easiest for you to set up to get it to compile at all.  You should not need MSVC to compile his plugin; if you really want performance you may as well get 2002 because um that's what he optimizes it for?

**Comment by vgturtle127** [comment]
@cxd4 True. What I don't get is he clearly has updated in the last few months, why not include a makefile or a solution file? I think I may create an issue on SourceForge.

**Comment by cxd4** [comment]
It wouldn't load on Visual Studio, unless you have VC .NET installed or try to auto-convert the proj.

Why not just build with MinGW?
* Takes between 50 MB to 1 GB of disk space, instead of 5 to 50 GB like Visual Studio does.
* Easily double-click a batch file to build from two simple commands to compile angrylion's two .CPP files.
* Doesn't use any background resources in RAM like keeping Visual Studio open.

**Comment by vgturtle127** [comment]
@cxd4 I will try that. I think I will still create an issue and ask him if he will change compilers or something. I will probably ask him why he hasn't uploaded binaries of any of the latest commits too, since the uploads just kind of stopped all of the sudden. :confused:

**Comment by JunielKatarn** [comment]
> Takes between 40 MB to 1 GB of disk space, instead of 5 to 50 GB like Visual Studio does.
Easy double-click a batch to build from two simple commands to compile angrylion's two .CPP files.
Doesn't use any background resources in RAM like keeping Visual Studio open.

A couple of clarifications:
* Visual Studio takes between 5 and 12 GB. Large, but not that large.
* No need to open the IDE at all. You can use the `MSBuild` command, which is what VS uses in the background.

I'd choose a compiler toolset in terms of the performance of its generated binaries (LIB, DLL, EXE).

**Comment by weinerschnitzel** [comment]
I think the 64DD cartridge conversions should receive extensive testing as well as those listed.

**Comment by cxd4** [comment]
>Visual Studio takes between 5 and 12 GB.

If you just use WE and/or Control Panel as a citation, then yes.  But on a system-wide basis, it has many (usually highly re-duplicated) sub-dependencies installations you can't really stay informed of.

Also, multiple installations of GCC toolkits are very consolidated.  Multiple VS are quite the inverse.

>I'd choose a compiler toolset in terms of the performance of its generated binaries (LIB, DLL, EXE).

Then clearly MSVC isn't the right choice for this, as anyone with experience must know.

Either way, if performance was on vgturtle127's list of "testing requirements", I'd still have suggested either GCC or, to match what angrylion optimizes for, VC2002.  2010+ is just going to make this slower.

**Comment by LegendOfDragoon** [comment]
>Either way, if performance was on vgturtle127's list of test requirements, I'd still have suggested either GCC or, to match what angrylion optimizes for, VC2002. 2010+ is just going to make this slower.

Isn't the performance better for Angrylion's plugin when using VC 2010+, compared to GCC?

**Comment by cxd4** [comment]
Haven't I explained this subject to you before?

Instruction cache size is a hardware limit.  angrylion even stated plainly himself that the balance of in-line expansion for the functions is specifically biased to MSVC 2002, so if you just write a GCC script that does -O3 and _INLINE all the things_ then **of course it's going to run slower**!

**Comment by cxd4** [comment]
That being said, I do genuinely appreciate you raising the question.  It's better than nodding what I say off like you agree with it when you don't, but I thought for sure I've discussed this before.

If you can post a script of the command you use to compile upstream plugin with MinGW for Windows then I can run a test on that and look at what optimizations are breaking the performance for your build.

**Comment by LegendOfDragoon** [comment]
I was genuinely curious if things have changed. I haven't been up to date with Angrylion's code. Last one I tried was r94, I think. Also, I haven't upgraded MinGW to the latest version. It's been a while since I tested Angrylion's with MinGW, I'll have to try it again sometime.

**Comment by cxd4** [comment]
I also have not upgraded myself.  GCC typically has more rocky roads with speed regressions vs. speed improvements, so I "can't wait" to see what the latest GCC compiles angrylion and rsp like.

I anticipate that due to the special case of VC2002-biased function expansion allocations, GCC would be better off compiling his plugin with either -O2, -O1 or even -Os:  not -O3.  Full optimizations with -O3 is practically going to overrride all his INLINE/NOINLINE advice and make the MinGW build way, way slower.

Should probably take my own advice, too.  As far as my OpenGL fork is concerned I have done -O2, -Os and -O3 out of intuitive understanding rather than actually comparing all of them for speed.  (In the short term this is a mistake of mine, but in the long term I think for the end result my settings are correct.)

Lastly, I recall observing that the VI compiles faster with MSVC than with GCC, but the RDP compiles faster with GCC than with MSVC.  That, again, is probably more of a coincidence.

**Comment by vgturtle127** [comment]
MinGW looks like to the one best suited here, I don't think Cygwin is appropriate here.

Cygwin has more overhead and less performance for some reason. I have never tried to figure out why.

Not sure about other compilers like GCC or Clang or something, but I'm waiting for Angrylion to reply to my issue to see if he can make quick work of all of this. There could be many reasons that he doesn't have a makefile or solution file, but I think regardless we can figure something out soon. :smile:

**Comment by cxd4** [comment]
I'm sure he has his reasons.  I don't even like supplying make files/scripts myself, honestly.

Is the time lapse between the latest commits by angrylion not being implemented into the OpenGL fork on EmuTalk the only reason you're disinclined to test with that?

**Comment by vgturtle127** [comment]
@weinerschnitzel Oh yeah! I forgot about Nintendo 64 Disk Drive support. I'm adding info about that above as soon as I am done with this comment! :smile: 

Thanks, these are great suggestions guys! Keep it up. :smiley: 

@LegendOfDragoon @cxd4 I had never even thought about the compiler's effect on speed in regards to my testing. I'm glad you guys brought it up, it is definitely something to think about. :thought_balloon:

**Comment by cxd4** [comment]
Yeah, that's something of a paradox, because the balance of FORCEINLINE/NOINLINE/INLINE in the plugin code is designed to optimize an inferior compiler version's fallback rules on in-line expansion.

Therefore, the better your compiler is at optimizing things for you, the slower that plugin gets.

**Comment by JunielKatarn** [comment]
> Then in that case, clearly MSVC isn't the right choice for this.

@cxd4 I'm interested in performance optimizations. How are you comparing toolchains? Is there any particular profiling techinque or tool for this?

**Comment by cxd4** [comment]
Well, on my current Athlon CPU I use AMD CodeXL to benchmark, but only when I "have" to.

That's a method I try to avoid because it's time-consuming.
But my options in order of preference are:

1.  the FPS and/or VI/s counters in emulators like Project64, which show emulation speed
2.  side-by-side comparison of the assembly language generated by both GCC & MSVC
3.  60-second benchmark profiles of all function times/latencies in e.g. AMD CodeXL

Generally the first is enough, unless I am committing micro-optimizations.  Project64's VI/s counter will usually be distinguishable enough between the MSVC and GCC builds or other major changes.

The second isn't really that hard to do and can sometimes be even more reliable than using professional software like CodeXL to profile your software for you, if you're familiar with a) instruction latency charts for different x86 CPU brands/versions b) instruction re-ordering rules for certain target CPUs to optimize for c) just to know the exact algorithm the compiler is translating the C to and why CodeXL says it's slower/faster (even if, in some contingencies, maybe it isn't).

**Comment by vgturtle127** [comment]
So I talked to Angrylion, and here is what he said:

I use MSVC 2003. Some time ago I tested a few compilers, and it compiled fine in Visual C++ 6.0, Visual C++ 2010 and in MinGW 4.6.3 (the last being the package that I downloaded from mamedev.org). This could have theoretically changed, but it's unlikely. Now compiling it in MSVC 2003 was discussed before on Project 64 forums and is easy: create an empty Win32 DLL project, add the 5 source files, make sure that the compiler sees Win32 and DirectDraw headers and that the linker sees user32.lib, ddraw.lib and dxguid.lib. I uploaded the project files in r98. Make sure to change the configuration to "Release" if you want to use them. Please tell me when I can close this ticket.

I haven't posted any binaries on EmuTalk, so you'd better ask cxd4 who did. cxd4 and I do not constitute a team. The builds he offered were based on his fork with his own numerous modifications and changes, for which I cannot be held accountable.

(After looking, HatCat/Iconoclast posted the binaries, not Angrylion or cxd4. So my thoughts are that those builds are not Angrylion's version, they are HatCat's fork.)

Angrylion posted the build files for MSVC 2003 (which apparently work with Visual Studio) in release r98. You can find that here:

https://sourceforge.net/p/angrylions-stuff/code/98/tree/trunk/mylittle-nocomment/

Would somebody try to compile it for me on Visual Studio? I only have about 175MB of space left on my hard drive! :fearful: 

I am going to separate the two builds from the list and add them separately. Thanks to Angrylion for his help!

**Comment by cxd4** [comment]
Creating VS project files is easy to do, but loading that one will either require going through the VS conversion wizard or installing MSVC ... 2003 (always thought I remembered reading 2002) or so.

As for your low hard disk space issue, I don't see why you're willing to go through all of that sort of trouble when you could just use a fairly recent build that was already released--either the one angrylion linked somewhere in one of his ticket discussions or the OpenGL fork by me.  I assume you're just trying to run accuracy tests on ROMs, right, not speed?  So why is it that important?

**Comment by vgturtle127** [comment]
@cxd4 Angrylion has never posted binaries as far as I know, so it would be important to have his build separate from HatCat's/Iconoclast's build, and your build as well.

Speaking of which, where can I find your build? I didn't know your build was separate from the other two.

In regards to your last statement, I go through as much trouble as I can. :smile: 

First and foremost, I want to test pretty much everything. I had no idea the Angrylion RDP plugin had two forks, so I want to test them both against the original.

I am trying to be as thorough as I can the first time, for completeness. I know we mentioned forks were kind of not worth testing, but I think many people think that the HatCat/Iconoclast build is the Angrylion build, and that your fork was a part of HatCat's/Iconoclast's build. I aim to clear up any confusion by not only listing them separately, but testing them separately as well. This will also bring to light any real world differences, and thus make the testing more complete.

**Comment by vgturtle127** [comment]
I updated the plugins and added info about N64DD. :smile:

**Comment by cxd4** [comment]
Whoever this HatCat/Iconoclast bloke is you keep referring to, he sounds like a whimpy-ass bitch.  Why isn't he updating his OpenGL fork as frequently anymore? :P 

I get what you're saying, but why file an issue on angrylion's SF tracker about the absence of a makefile/solution for building with VC, if you don't even have the hard disk space to attempt building it yourself anyway?  I mean, if you can't even use the solution file he just provided you with due to insufficient hard disk space (I recommend cleaning files, btw), then it's kind of like you may as well just close/resolve your own issue on his tracker.

Yes we could build it ourselves, but me I am already busy with my own fork.  I would just build that and release a next version when I feel like it.  You can always free up disk space or use one of the pre-compiled binary releases.

**Comment by vgturtle127** [comment]
@cxd4 Well, because I thought people would want to build the original for themselves, unmodified from Angrylion's source (since he provided much of the MAME/MESS/UME RDP and RSP fixes). The two builds (Angrylion's and HatCat's/Iconoclast's) aren't source identical from what I can tell, even by comparing the .DLL without source. EDIT: check below for additional info

BTW, you keep saying I can build your fork, but I can't find it anywhere. Are you sure you aren't confusing your RSP fork? We are talking about the RDP plugin by Angrylion. You forked that somewhere? I'm just making sure we understand each other... :blush: 

I am in the process of cleaning my hard drive of non-N64 related stuff now. Most of it is research, source code, and dumps by myself. So I have to save that. It isn't exclusively trash and Windows crap. I just ran Disk Cleanup and saved about 14MB. :laughing: 

When you say pre-compiled binary releases, what do you mean? I have HatCat's latest build, but Angrylion has never posted binaries. So that means the only available pre-compiled binary releases are HatCat's/Iconoclast's or modified builds of Angrylion. I have never seen any binaries posted that were actually Angrylion's unmodified plugin. I have the source and I will compile and test Angrylion's beside HatCat's. When you can give me the location to source or binaries for your plugin, I will add that to the list as well.

EDIT: I found out the difference between Angrylion's build and HatCat's build. Angrylion exclusively used DirectDraw for his plugin. HatCat apparently ported it to OpenGL, and also added support for Linux N64 emulators (Mupen64, essentially). He also said he "likes to staticize the C code (which includes, but is not limited to, SIMD vectorization. For example, MMX)". So that too.

You can read his full statement here: http://www.emutalk.net/threads/55481-angrylion-s-Per-Pixel-RDP-with-OpenGL

Project64 forum member Haxatax posted a build of Angrylion's plugin, but that is an older version and changes the default resolution. So it is different than the default build again. I added all the updated info to my original post, and also added the architecture of each plugin. I also updated most of the information and cleaned up a little.

**Comment by cxd4** [comment]
Yep, ported it to OpenGL once I realized there were several permanent issues with DirectDraw.

>Most of it is research, source code, .... So I have to save that.

Well if it's source code, why keep it on your hard disk when it's full?

Couldn't you just push it to Git repositories?  That's the best way to keep your code backed up.

Same thing with the other things you said (research, dumps) ... just back it up online to the cloud if it's nothing you're concerned over your grandma seeing, then the disk space can be freed.

**Comment by vgturtle127** [comment]
@cxd4 That is what I'm doing now, pushing most of it to my Git repos. However, the dumps are copyrighted materials, so I can't upload them anywhere. I can store them on my external drive though, which is what I'm also doing now.

I don't like to back up to the cloud anyway, I don't trust most companies that have cloud backup services. If I lost anything, I would go crazy. Some of the stuff I have is either super rare stuff that I have dumped, or some of it is stuff I created for whatever reason. Sorry for being vague, but I don't want to disclose any information until it is ready for release. :smile:

**Comment by vgturtle127** [comment]
Does anybody know if this is Bsmiles RSP?

https://github.com/mupen64plus/mupen64plus-rsp-hle/tree/bsmiles32/master

Or is this it?

https://github.com/bsmiles32/mupen64plus-rsp-hle

**Comment by weinerschnitzel** [comment]
https://github.com/mupen64plus/mupen64plus-rsp-hle/

It looks like this is the current repository. It also appears they are talking about splitting off the m64p plugin spec from the HLE RSP core. Good for us, as far as having a zilmar spec plugin based on the same codebase.

**Comment by vgturtle127** [comment]
Oh, I thought that one was @cxd4 's RSP plugin. It;s hard to keep track. :blush: 

Thanks, @weinerschnitzel !

**Comment by vgturtle127** [comment]
So I think I am going to start the testing soon. What I'd like to do is see if anybody intends on releasing new stable versions of their plugins (or of Project64, for that matter). I will ping everybody who I think should be made aware.

@project64 @Azimer @cxd4 @purplemarshmallow @toehead2001 @AmbientMalice @LegendOfDragoon @BigheadSMZ @theboy181 @tony971 @dsx- @death-droid @oddMLan @ExtremeDude2 @LuigiBlood @JunielKatarn @MELERIX @Lithium64 @samuelyuan @Frank-74 

Feel free to ping anybody else who might have something to add, say, etc.

I think I will create a repository to keep all the testing documentation in.

I will post a link when it is up. :smiley: 

I'm excited. :open_mouth:

**Comment by ExtremeDude2** [comment]
I feel so loved!

**Comment by vgturtle127** [comment]
@ExtremeDude2 :laughing: 

Hey, you are important!

**Comment by theboy181** [comment]
What are you testing for?

**Comment by theboy181** [comment]
What army AIO's pluging included on this list.  Hehehe.

**Comment by toehead2001** [comment]
> Why is there no obscure testing with closed source plugins? These often lend clues to the possibilities and potentials.

If you want clues to the possibilities and potentials, run the game on the original N64 hardware. :smirk:

**Comment by theboy181** [comment]
And so I do.

**Comment by weinerschnitzel** [comment]
I recently forked Bsmiles RSP and made minimal changes to build it for Zilmar Spec emulators. I essentially made the same changes I did for Surreal64. @LegendOfDragoon I don't think I made any changes that you may consider too exciting, but I have it up now. @cxd4 I could use some corrections from you to meet rcp64 standards if you want to have a look-see.

https://github.com/weinerschnitzel/mupen64plus-rsp-hle/tree/PJ64_SPEC

**Comment by LegendOfDragoon** [comment]
@weinerschnitzel thanks for creating the fork! I will test it out tonight. Do you still go on IRC? Would be nice to chat about certain stuff.

**Comment by cxd4** [comment]
@weinerschnitzel : Yes, in this file:
https://github.com/weinerschnitzel/mupen64plus-rsp-hle/blob/PJ64_SPEC/src/Rsp.h#L43

This would break zilmar-spec plugin support on 64-bit non-Windows operating systems.

unsigned long* (any RCP register) is unstable because `long` is a 64-bit data type on all 64-bit operating systems we'll ever encounter, with the sole exception of 64-bit Windows which is dumb as hell and should be ignored.

**Comment by cxd4** [comment]
Other than that, I think your version is basically 100% binary-compatible with zilmar plugin spec.

You can say `unsigned long Cycles` instead of `uint32_t Cycles` too, if you want.  You don't have to force it to 32-bit; you'll just get extra precision for 64-bit cycle counts on CPU hardware with native support for it.  The only thing it will break is zilmar's stubborn decision to force it down to 32-bit-only which I see as entirely unwarranted.

**Comment by LegendOfDragoon** [comment]
@weinerschnitzel , looks like your fork also has the issue with Body Harvest. I wonder which HLE function needs to be fixed.

**Comment by weinerschnitzel** [comment]
Thanks for the quick response! :) I'll see about borrowing your my_types.h and Rsp.h from rcp64 to make it more consistent. I wanted to get something together quickly, so there are some things that are a little sloppy.

**Comment by cxd4** [comment]
You're also welcome to pull request any ideas for my `rcp64` repository.

I'm a bit indecisive about a few of my portability standardization changes to zilmar's plugin spec headers, but I try to be as decisive as I can.  I'm sure others will have ideas besides mine that can improve it.  (No changing every single instance of `int` to `int32` is NOT one of them.)

**Comment by vgturtle127** [comment]
@theboy181 I am testing some closed-source plugins. What do you mean?

@weinerschnitzel Dang dude! That was a pretty quick conversion! Cool! I have added that to the list and I will test it when it is working. Or is it working? Or have you not tested it yet?

@project64 Is 2.3 going to be out relatively soon? I want to start the testing ASAP, but if a new stable is coming soon I don't want to start right before it comes out.

@gonetz Do you happen to have an ETA on the next stable version of GlideN64? If not, that is fine. Also, is 64-bit Project64 support coming? Just want to make sure I don't start testing right before releases...

@olivieryuyu @phire @Catacea3691 @lioncash @SwooshyCueb @schibo 

Pinging some more people who have contributed greatly to N64 emulation.

**Comment by weinerschnitzel** [comment]
I only tested it with PJ64 1.6. I still need to fix somethings with it so Mupen64 and 1964 can use it. It's also missing config and about dialogs. I need to build a new version of PJ64 2.x to test that as well.

**Comment by vgturtle127** [comment]
@weinerschnitzel Alright, thanks so much dude. Awesome work!

**Comment by tony971** [comment]
Since it's pretty safe to assume that you're not compiling anything, I thought I'd give you everything that I compile myself. So here's the latest versions of:

* Project64
* GLideN64
* Angrylion's unmodified plugin
* Aziaudio
* HatCat's RSP

And for future reference, HatCat/Iconoclast is @cxd4. . So when he says "my code," he means anything made by HatCat.

https://www.dropbox.com/s/g6qj0j2whro14jp/Project64.7z?dl=0

**Comment by project64** [comment]
I am making to many changes to the source for any release to be coming out soon

**Comment by theboy181** [comment]
@tony971

Interested in Angrylion's modified code. What has been modified?

**Comment by tony971** [comment]
It actually says "unmodified." Meaning it's angrylion's original fork. Not HatCat's.

**Comment by vgturtle127** [comment]
@tony971 

> safe to assume you're not compiling anything

Not entirely true, I just can't install Visual Studio since it would fill my SSD. I have MinGW installed, but that has it's own issues.

> HatCat/Iconoclast is cxd4

WHAT?! I feel like an idiot, but at the same time, how was I suppose to know? Nobody (including cxd4) said anything! :laughing: 

@project64 OK, thanks for the info. I like that you're cleaning so much up though, it already looks way better. :smile: 

@theboy181 @tony971 I tried to play some games using Angrylion's RDP the other day, but performance is not really playable when there is 3D rendering. In menus and stuff it is OK, but anything else is 10-25 frames per second and changes pretty drastically. It should make testing it fun though, because maybe it isn't like that with all games?

**Comment by gonetz** [comment]
> Do you happen to have an ETA on the next stable version of GlideN64? 

Not yet.

**Comment by vgturtle127** [comment]
@gonetz Alright, thanks buddy! By the way, those improvements to 2D are so awesome! I thought I noticed something off when I was playing some games yesterday. :smile: 

What's weird is that Angylion's RDP breaks those 2D images as well. Not sure why, though. :hushed:

**Comment by vgturtle127** [comment]
@tony971 Thanks for compiling that stuff for me, but what commit did you compile Project64? Full screen is completely broken. It corrupts graphics, and when activated it doesn't even give me real fullscreen. It just makes the window huge...

**Comment by LegendOfDragoon** [comment]
>It should make testing it fun though, because maybe it isn't like that with all games?

Some games have ok speeds.

>What's weird is that Angylion's RDP breaks those 2D images as well. Not sure why, though. :hushed:

Can you explain?

**Comment by tony971** [comment]
Full screen is completely broken. It corrupts graphics, and when activated it doesn't even give me real fullscreen. It just makes the window huge...

Huh... I thought that was a local issue. Apparently not. And I compiled 55bc654 but I've noticed that issue on earlier builds. Something in config or cache files fixes it, though. I only experience the issue on fresh builds. When placed into my normal folder with existing configs/caches, fullscreen renders fine.

**Comment by Frank-74** [comment]
I though Angrylion's 'my little comment' plugin was 640x480 resolution only?  That one is about 1024x768?

**Comment by tony971** [comment]
@vgturtle127 do you experience the issue when you start in windowed mode and switch to fullscreen after emulation is started?

**Comment by vgturtle127** [comment]
@LegendOfDragoon I mean that even Angrylion's RDP (unmodified or cxd4's build) has issues with 2D that other plugins have, like Glide64, GlideN64, Jabo's, etc. The title screen of Mario Kart is a good example.

@tony971 I tried deleting the cache files, that didn't fix it either. Not sure, maybe a regression? I do experience the issue starting in windowed mode. I always start in windowed mode.

@Frank-74 What do you mean "that one"?

**Comment by cxd4** [comment]
@Frank-74 angrylion's upstream code is fixed to 1024x768.

My fork of it had it changed to 640x480 since the very beginning.  As angrylion himself said in the commit, the choice of 1024x768 was just personal taste.  I see nothing other than 640x480 to be a logical default configration however.

@vgturtle127 : fairly sure angrylion almost outright told you I was him in his reply.

**Comment by vgturtle127** [comment]
@cxd4 No, he said you guys weren't a team, and that he didn't know about anything else that was going on besides the work he was doing.

And to be fair, I had no reason to think any different. Completely different usernames that had nothing to do with each other as far as I could tell. :blush: 

Maybe I missed something? Did everybody else know this the whole time?

**Comment by tony971** [comment]
He's also BatCat and DrCat in IRC. He's been many cats. Iconoclast threw me through a loop for a second, though.

**Comment by Frank-74** [comment]
@vgturtle127 I meant the Angrylion plugin @tony971 included in his archive above.

**Comment by cxd4** [comment]
>Maybe I missed something?

_"cxd4 and I do not constitute a team. The builds he offered were based on his fork with his own numerous modifications and changes,"_

In response to you asking about "his" builds on EmuTalk, he told you cxd4's builds (meaning, ask cxd4 who's builds they are) are his own fork etc. etc.  He basically did tell you I was him; you just didn't read deep enough into his reply.

I am many more names that I don't care to remember here. :)

**Comment by vgturtle127** [comment]
@tony971 Yeah, what is an Iconoclast? It sounds cool, at least...

@Frank-74 Tony included the unmodified Angrylion plugin. That issue occurs on every version of the plugin, though.

@cxd4 I don't know, I got the vibe that he didn't know those other names you went by. Whatever, I know now. :laughing:

**Comment by weinerschnitzel** [comment]
Yesterday, I made the functional changes I wanted to for my zilmar spec fork of bsmiles' RSP plugin. I think it is functionally where I'd like it, and I made a post about it on EmuTalk to receive feedback. I've unsubscribed from the party thread. Contact me on ET or create an issue on my fork's branch.

@vgturtle127 Feel free to ping me when you have completed testing for this ticket, or are starting to publish your results.

**Comment by vgturtle127** [comment]
@weinerschnitzel Alright, thanks dude.

@Azimer Is 0.70 going to come out soon? Just curious.

**Comment by fallaha56** [comment]
@vgturtle127 reading this, liking the concept++

just wondering, would it be worth making all this into a wiki? am sure lots of folk would love to chip in and help with this mammoth but worthy task...

**Comment by vgturtle127** [comment]
@fallaha56 I will be publishing all the data on a GitHub repository, so if anybody wants to contribute they can. A Wiki is sort of the direction I am going, except I will just be posting all my data in markdown. It is easily converted to Wiki format though.

I'm not sure if anybody intends on "chipping in", but I think in general people will wait and see the results first. After that, people may use my information as supplementary data to fix things and make improvements. At least, that's what I'm assuming.

**Comment by vgturtle127** [comment]
@schibo 

I could use your help with 1964 information and plugins!

**Comment by vgturtle127** [comment]
Hey, guys. Sorry if it seems like nothing is going on.

I have acquired a large amount of plugins, and I am testing to make sure they all work in Project64.

I am also trying to gather N64 documentation. I have a pretty good collection now. :smirk: 

A friend of mine just released some stuff to me, an RSP microcode assembler and compiler. Pretty interesting stuff. I will have a GitHub repository link soon. :smile: 

I also had the same friend give me a ton of N64 documents and source not available to the public, so I hope that these will help me in better understanding the hardware. :computer: 

I have updated the original post with new plugins and more information, so you can read that and check my work in the meantime. :wink: 

Hopefully, I will be able to start testing in the near future. All I really have left to do is try and make more space so I can extract all my ROM's. I hope to have that figured out soon as well.

**Comment by LegendOfDragoon** [comment]
@vgturtle127, I think you're trying to test too many plugins. There is absolutely no point in testing plugins such as Shunyuan's Soft Graphic plugin. WDC, Stunt Racer, and Gauntlet don't work, and a few games have graphical glitches such as SD Hiryuu and Mortal Kombat 4. Not much point in testing Azimer's LLE audio plugin (0.50 beta 1) either, but if you insist on testing it, I'd only try it in problematic games such as Top Gear Rally, Body Harvest, Indiana Jones, Star Wars Battle For Naboo, and Star Wars Rogue Squadron. Also, I'm pretty sure Schibo's audio is closed source, if not, then I'd like to see it :smile: .

**Comment by vgturtle127** [comment]
@LegendOfDragoon Yeah, I tried to ping Schibo but I'm not sure he is very active on GitHub.

I really only wanted to test Azimer's LLE audio because his HLE is so good, I figured his LLE plugin might work great with Angrylion's RDP or something. I have never used it, and I didn't know it existed. It was interesting to me, is what I'm saying.

Why does everybody say Shunyuan's plugins aren't worth testing? Isn't anybody curious? I am. :laughing: 

I can axe them for testing, but the point is to find each plugins pros and cons, and to find bugs and inaccurate behaviors. Basically, an N64 emulation compatibility bible. But I understand certain plugins (like some input plugins or ancient graphics plugins) should be axed.

If you guys really feel like a plugin isn't worth testing, I'd really like to hear it. The more I cross of the list, the easier and faster testing will be. :smile: 

I'd like to hear from more people about plugins they think should or shouldn't be tested. :smiley:

**Comment by LegendOfDragoon** [comment]
>I really only wanted to test Azimer's LLE audio because his HLE is so good, I figured his LLE plugin might work great with Angrylion's RDP or something. I have never used it, and I didn't know it existed. It was interesting to me, is what I'm saying.

The thing is, you can use LLE for any audio plugin, afaik. HLE is just an option, for some plugins.

>Why does everybody say Shunyuan's plugins aren't worth testing? Isn't anybody curious? I am. :laughing:

His audio plugin is "usable" but there's no practical reason to use his graphics plugin. His graphics plugin is based off of outdated code and he used an unsuitable compiler, so the plugin's extra slow. His audio plugin is actually decent in games such as Mortal Kombat 4 and Cruise N World, so if you are curious about it, you're free to test that (if you can get past the issue that it crashes the emulator often).

Also, I couldn't get purplemarshmallow's Z64 build to work. Does it work for you?

**Comment by vgturtle127** [comment]
@LegendOfDragoon Oh, so the checkbox works on any plugin? So the plugin has to have both methods in order for it to work? That is either really cool or totally confusing.

What do you mean? Do you mean you can't get it to load into Project64, or do you mean it crashes or something before the emulator does anything?

On the latest build of Project64, it shows up in my plugins fine.

**Comment by LegendOfDragoon** [comment]
>@LegendOfDragoon Oh, so the checkbox works on any plugin? So the plugin has to have both methods in order for it to work? That is either really cool or totally confusing.

If an audio plugin supports HLE, it supports both HLE and LLE audio. Plugins like Jabo's audio only support LLE.

>What do you mean? Do you mean you can't get it to load into Project64, or do you mean it crashes or something before the emulator does anything?

Crashes / doesn't work when booting a game.

**Comment by vgturtle127** [comment]
@LegendOfDragoon Oh OK, good to know.

What is the error message when you boot the game? I get the game to boot, but most of the time it shows a black screen...Which is weird because some games were working a while back?

**Comment by vgturtle127** [comment]
@project64 Would you say the interpreter or the recompiler is more accurate? Are both in terms of CPU features complete?

**Comment by project64** [comment]
@vgturtle127 interpreter is more accurate, tho they should be the same and when they are different then I fix the recompiler to match. The recompiler just has a lot more complications like self modifying code, and that it actually compiles code so there is more chances of something going wrong.

**Comment by vgturtle127** [comment]
@project64 OK, just curious about your opinion. :smile: 

I actually haven't tried the interpreter for most games yet, so that is on my TODO list. :smile: 

By the way guys, all my work will be put into this repository:

https://github.com/vgturtle127/Project64-work

This will include the data I collect during the audit, in-depth documents about specific games, their quirks, the N64 hardware, etc. I am doing most of the work locally for a little while, just so I don't get people trying to do pull requests and stuff for data I already have. :smile: 

Thanks for your patience, guys. Almost there. :smile:

**Comment by LegendOfDragoon** [comment]
>I actually haven't tried the interpreter for most games yet, so that is on my TODO list.

The most useful way to test both is to set cpu core style to synchronize cores (which requires Enable Debugger to be on). That way, you can see whether the accuracy for both interpreter and recompiler is the same. Keep in mind that different recompiler options can have an effect on the results. Keep in mind it requires Fixed audio timing OR no audio, in order to give proper results. Also, it's very resource intensive on your cpu, so be sure to use a fast video plugin and preferably HLE audio as well.

**Comment by vgturtle127** [comment]
@LegendOfDragoon I can't find the "synchronize cores" option. I have the debugger enabled, but most of the debugger options crash Project64...

**Comment by rafaelsantino** [comment]
> * [x]  UltraHLE video plugin (version?) [closed-source?] {x86}

i tested that plugin already, but heres a tip for make it run: run project64 in administrator, since ultrahle saves its settings in registry
